netroute_bsd.go

```diff
@@ -13,137 +13,377 @@
 package netroute
 
 import (
+	"errors"
 	"fmt"
 	"net"
-	"sort"
+	"os"
+	"sync/atomic"
 	"syscall"
+	"time"
 
 	"github.com/google/gopacket/routing"
 	"golang.org/x/net/route"
+	"golang.org/x/sys/unix"
 )
 
-func toIPAddr(a route.Addr) (net.IP, error) {
+const (
+	RTF_IFSCOPE     = 0x1000000
+	ROUTE_MSGFILTER = 1
+)
+
+type bsdRouter struct {
+	id  uintptr
+	seq atomic.Uint32
+}
+
+// toIPAddr converts a route.Addr to a net.IP.
+// Returns nil if the address type is not recognized.
+func toIPAddr(a route.Addr) net.IP {
 	switch t := a.(type) {
 	case *route.Inet4Addr:
-		return t.IP[:], nil
+		return t.IP[:]
 	case *route.Inet6Addr:
-		return t.IP[:], nil
+		return t.IP[:]
 	default:
-		return net.IP{}, fmt.Errorf("unknown family: %v", t)
+		return nil
 	}
 }
 
-// selected BSD Route flags.
-const (
-	RTF_UP        = 0x1
-	RTF_GATEWAY   = 0x2
-	RTF_HOST      = 0x4
-	RTF_REJECT    = 0x8
-	RTF_DYNAMIC   = 0x10
-	RTF_MODIFIED  = 0x20
-	RTF_STATIC    = 0x800
-	RTF_BLACKHOLE = 0x1000
-	RTF_LOCAL     = 0x200000
-	RTF_BROADCAST = 0x400000
-	RTF_MULTICAST = 0x800000
-)
+// toRouteAddr takes a net.IP and returns the corresponding route.Addr.
+// Returns nil if the IP is empty or has an invalid length.
+// IPv4 addresses are converted to route.Inet4Addr and IPv6 to route.Inet6Addr.
+func toRouteAddr(ip net.IP) route.Addr {
+	if len(ip) == 0 {
+		return nil
+	}
 
-func New() (routing.Router, error) {
-	rtr := &router{}
-	rtr.ifaces = make(map[int]net.Interface)
-	rtr.addrs = make(map[int]ipAddrs)
-	tab, err := route.FetchRIB(syscall.AF_UNSPEC, route.RIBTypeRoute, 0)
-	if err != nil {
-		return nil, err
+	if len(ip) != net.IPv4len && len(ip) != net.IPv6len {
+		return nil
 	}
-	msgs, err := route.ParseRIB(route.RIBTypeRoute, tab)
-	if err != nil {
-		return nil, err
+	if p4 := ip.To4(); len(p4) == net.IPv4len {
+		return &route.Inet4Addr{IP: [4]byte(p4)}
 	}
-	var ipn *net.IPNet
-	for _, msg := range msgs {
-		m := msg.(*route.RouteMessage)
-		// We ignore the error (m.Err) here. It's not clear what this error actually means,
-		// and it makes us miss routes that _should_ be included.
-		routeInfo := new(rtInfo)
-
-		if m.Version < 3 || m.Version > 5 {
-			return nil, fmt.Errorf("unexpected RIB message version: %d", m.Version)
-		}
-		if m.Type != 4 /* RTM_GET */ {
-			return nil, fmt.Errorf("unexpected RIB message type: %d", m.Type)
-		}
+	return &route.Inet6Addr{IP: [16]byte(ip)}
+}
 
-		if m.Flags&RTF_UP == 0 ||
-			m.Flags&(RTF_REJECT|RTF_BLACKHOLE) != 0 {
+// ipToIfIndex takes an IP and returns the index of the interface with the given IP assigned.
+// Returns an error if no interface is found with the specified IP address.
+func ipToIfIndex(ip net.IP) (int, error) {
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return -1, fmt.Errorf("failed to get interfaces: %s", err)
+	}
+	for _, iface := range ifaces {
+		addrs, err := iface.Addrs()
+		if err != nil {
 			continue
 		}
```

Comment:
Should these errors be ignored? (I'm not sure)
Maybe it's fine so long as we get _any_ interface, but if not, we could maybe aggregate them.
```go
// outter scope
var errs []error
// inner scope
if err != nil {
    errs = append(errs, err)
    continue
}
//outter scope
if err := errors.Join(errs...); err != nil { 
    return err
}`
```

Reply:
I think aggregating makes sense

Reply:
Actually upon further review we can ignore, we only care to find ifaces with a specific addrs if a particular iface fails to return address, it really don't matter. The assumption is made that it is most likely not the interface we want.